text/0000-clarity-api-description.md

```diff
@@ -0,0 +1,77 @@
+- Start Date: 2015-06-18
+- RFC PR: (leave this empty)
+- Refract Issue: (leave this empty)
+
+# Summary
+
+Provide clarity and additional elements/properites in API Description namespace.
+
+# Motivation
+
+The purpose of this document is to make it easier for newcomers to understand the format and build tools around it.
+
+# Detailed design
+
+## Combine Namespaces
+
+We should combine the API Description namespace with the resource namespace.
+Right now they are separate and it would be clearer if they were not.
+
+## Elements in Attributes
+
+Refract allows for including elements within the attributes section of an element.
+Even though it’s possible, it may be helpful to avoid this in such an important namespace.
+Reason being, this will be one of the first places people interact with Refract, and to use complex features may be too much.
+
+## HREF Variables
+
+We should call them “hrefVariables” everywhere instead of parameters in places (specifically transitions).
+I think this will be more consistent and easier to understand for newcomers.
+This will make more sense in how we handle inheritance as well, because parameters inheriting from hrefVariables does not seem optimal.
+
+## Defining Classes
+
+Swagger provides a way to define tags that can be applied to resources throughout the document.
+Refract has the “class” attribute that we can use the same way, we only need a way to define a class name and provide the description for it.
+Of course, this won’t carry over to API Blueprint, but that’s OK.
```

Comment:
Paraphrasing from @smizell & the [Swagger spec](https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md#tagObject): a resource may include a list of tags as a way of grouping items for tooling authors. These tags can contain a name, description, and external resource link. One resource may have multiple tags.

You can think of this as similar to, but more flexible, than our resource groups. A resource can only be in one resource group. Also, resource groups can create a hierarchy, while tags (as used in Swagger) cannot.

To me this is a generic API description feature. I believe that a flexible way to group resources would be beneficial to the adoption of a generic API description format.

I would propose that we either:

1. Create a hierarchical tag element for refract that references resources defined elsewhere in the document. Our resource groups would be implemented as this new element and we can support Swagger and other formats that use tags as well. For example:

 ```yaml
element: category
meta:
  class: ['api']
content:
  -
    element: category
    meta:
      class: ['resources']
    content:
      -
        element: resource
        meta:
          title: 'Resource 1'
        attributes:
          tags: ['tag1']
        content: ...
      -
        element: resource
        meta:
          title: 'Resource 2'
        attributes:
          tags: ['tag2']
        content: ...
  -
    element: category
    meta:
      class: ['tags']
    content:
      -
        element: tag
        meta:
          title: 'tag1'
          description: 'Some tag description'
        content:
          -
            element: tag
            meta:
              title: 'tag2'
              description: 'I am a nested tag'
          -
            element: externalRef
            content: 'http://example.com/'
 ```

 This is really powerful but could present some interesting problems, such as how the heck to render the documentation since a resource may be in multiple tags.

2. Leave things as-is and add a flat (i.e. non-hierarchical) tag element to support Swagger's use case. Tooling authors will need to support both resource groups and tags, but the tagging system is simplified. I think this is more explicit than re-using the `class` metadata. So far we have used the `class` metadata to differentiate the specific type of element (i.e. group of resources vs. group of data structures), not as an arbitrary grouping mechanism. **If** resources have a unique ID in the document then they should be easy to reference with a tag element.

 ```yaml
element: category
meta:
  class: ['api']
content:
-
  element: category
  meta:
    class: ['resourceGroup']
  content:
    -
      element: resource
      meta:
        id: some-document-unique-identifier
      content: ...
-
  element: category
  meta:
    class: ['tags']
  content:
    -
      element: tag
      meta:
        title: tag1
        description: I am a tag
        href: http://example.com/
      content: '#some-document-unique-identifier'
 ```

 This will be simpler to implement but maybe less powerful. What do you think?